src/db-transaction/db-transaction.service.ts

```diff
@@ -0,0 +1,40 @@
+import { Injectable, OnModuleDestroy } from '@nestjs/common';
+import { DataSource, EntityManager, QueryRunner } from 'typeorm';
+import { IsolationLevel } from 'typeorm/driver/types/IsolationLevel';
+
+@Injectable()
+export class DbTransactionService implements OnModuleDestroy {
+    private readonly queryRunnerSet: Set<QueryRunner>;
+    constructor(private readonly dataSource: DataSource) {
+        this.queryRunnerSet = new Set();
+    }
+
+    async execute<T>(
+        executable: (manager: EntityManager) => Promise<T>,
+        isolationLevel: IsolationLevel = 'SERIALIZABLE',
+    ): Promise<T> {
+        const queryRunner = this.dataSource.createQueryRunner();
+        this.queryRunnerSet.add(queryRunner);
+        await queryRunner.connect();
+        await queryRunner.startTransaction(isolationLevel);
+        try {
+            const result = await executable(queryRunner.manager);
+            await queryRunner.commitTransaction();
+            return result;
+        } catch (err) {
+            // since we have errors, rollback the changes we made
+            await queryRunner.rollbackTransaction();
+            throw err;
+        } finally {
+            // we need to release a queryRunner which was manually instantiated
+            await queryRunner.release();
+            this.queryRunnerSet.delete(queryRunner);
+        }
+    }
```

Comment:
If all suggested is true, such function needs no separate folder, it could be chunked in the common utils folder

Reply:
Okay